### ISSUE TITLE
Update our cre monospace font?

### ISSUE THREAD

**Issue by poire-z** [koreader/koreader#9211]
(I think I've seen this mentionned, in some issue or discussion, but can't find it.)

I find our mono font (DroidSansMono.ttf) a little too large vs most of the other serif fonts we may use with books (Noto Serif here):
![image](https://user-images.githubusercontent.com/24273478/173570631-59d9dd0b-e8d9-4319-9ba5-d90c8f968177.png)

Is it just me ? Or is this just doomed to happen with a monospace fixed-char-width that can only look bigger than a proportional font that gets each letter condensed/smaller ?
I don't often read "computer language/code" books , so it doesn't really hurt me - but when I do, it feels a bit ugly (and it's something I just can't fix by hacking on the crengine side :) also, we have no way of setting another font, dunno if we should have that, or if it's enough to find a nicer font).

(Some versions ago, Firefox on Windows switched from using the old/big Courier New to using Consolas, and it gives a noticable refreshment when looking at monotype parts.
![image](https://user-images.githubusercontent.com/24273478/173571767-28fc71fd-8a59-48b7-8b0d-448cf0f7c880.png)
https://bugzilla.mozilla.org/show_bug.cgi?id=713680
https://bugzilla.mozilla.org/show_bug.cgi?id=1607913
)

I have no idea if there are free candidates - and I'm not really interested looking around. Just opening this issue in case other people are, or already have ideas.
I have also no idea about this DroidSans' Unicode scripts/languages coverage.

**Comment by NiLuJe** [comment]
Consolas is my preferred choice, FWIW, but it's obviously non-free ;).

(I *think* there are free clones around, but I'm not sure they're all that great, and coverage will probably be an issue).

I *hate* all the Courier with a fiery passion (but there *are* good free candidates there, IIRC).

----

KoboUSBMS uses https://github.com/microsoft/cascadia-code (well, the NerdFont variant ;p), and I like it as an UI font, but I don't recall ever having tried it for large blocks of texts ;).

**Comment by mergen3107** [comment]
Consolas is my favorite as well, oi wei it isn’t free. 
i use in Notepad++ and starting a specific font size it becomes nice boldy and contrasty :D so pleasant to read my Matlab and Fortran codes :D

**Comment by offset-torque** [comment]
**I am approaching this from an UX perspective:**
If I am reading a programming book, I would like inline code to be easily distinguishable when surrounded by normal text. In poire-z's first example image, mono font is very similar to the serif. There is no glanceability. I missed that **object** keyword on the fourth line while reading and if it wasn't slightly bigger it would totally blend in.

I checked some programming books to see what is the general practice. Here are examples from Packt (top) and O'Reilly (bottom):

![monospace font](https://user-images.githubusercontent.com/84420582/173600498-8ee1c35e-7c56-4aaa-8df7-63eaaa244ace.png)

Monospace parts are easier to distinguish as you can see.
Do we really want to make monospace text to blend in to the rest of the text instead of standing out more ?
Is there a reason we are going against industry practices ?
Am I missing something obvious here ?

**Comment by poire-z** [comment]
> Is there a reason we are going against industry practices ?

Come on, you should know us by now. We're going against industry practices _just because we can_ !

Just joking :)
We're not going anywhere, I'm just opening this for discussion, and your input is good, don't accuse us of doing already the bad choice :)

> In poire-z's first example image, mono font is very similar to the serif. There is no glanceability. I missed that object keyword on the fourth line while reading and if it wasn't slightly bigger it would totally blend in.

See, you admit like me that our monospace font could be better.
I prefer the monospace font on your screenshot that on mine. On mine (current KOReader), it's both taller and larger than most (I think) normal font. In your screenshot, it is still wider (as expected with a monospace), but smaller in height - and this "smaller in height" is enough for me to make it look better.
(And I'm quite sure I'll be happy if our current font would have its size set 90% or 95%.)

**Comment by Frenzie** [comment]
> Is there a reason we are going against industry practices ?
Am I missing something obvious here ?

a. These are industry practices. Industry practices can be f@#$ awful!
b. The fact that this topic is about changing to our own and/or different industry practices? I don't really understand what point you're trying to make here.

**Comment by offset-torque** [comment]
@poire-z 
> We're not going anywhere, I'm just opening this for discussion, and your input is good, don't accuse us of doing already the bad choice :)

All is cool, really. No accusations. I commented in good faith. I am just curious this time :)

> On mine (current KOReader), it's both taller and larger than most (I think) normal font. In your screenshot, it is still wider (as expected with a monospace), but smaller in height - and this "smaller in height" is enough for me to make it look better.

I totally agree with this. Making it smaller is a better direction than bigger to add difference (and it looks nicer).

> (And I'm quite sure I'll be happy if our current font would have its size set 90% or 95%.)

Accurate. I tested before writing the first message and they look nearly the same at 10pt (noto) vs 9.5pt (droid). Probably 90% will look satisfactorily smaller.


@Frenzie 
> b. The fact that this topic is about (maybe) changing to our own and/or different industry practices? I don't really understand what point you're trying to make here.

No point, just trying to understand. I have used many programming books in the past (PHP, Arduino/Processing) and every one of them had a very distinguishable mono font (sometimes painfully). Depending on this information and my experience I always thought everyone prefers this.

Changes about this issue doesn't really affect me, I am not reading any books on my e-reader which require mono fonts. And programmers are knowledgeable enough to open an issue if this bothers them. I am just curious as I said.

---

Slightly related to the topic: There is this page where you can choose between programming fonts tournament style to find your favorite - [https://www.codingfont.com/](https://www.codingfont.com/)

**Comment by Frenzie** [comment]
> No point, just trying to understand.

Why we use these particular fonts is more or less explained here: <https://www.npr.org/sections/codeswitch/2014/08/03/337168933/-no-tofu-doesn-t-equate-to-no-problem-for-google-universal-typeface>
tl;dr they have a lot of characters.

I think nobody is super happy with them, but they're useful and they're not actively bad either. (They're designed to be bland and unoffensive.) Most types we might prefer will have plenty of tofu or licensing issues, so there's a chance we'd need it as fallback regardless.

But I'll reiterate that we're not going "against" anything. These are "industry standard" Adobe creations.

**Comment by poire-z** [comment]
@NiLuJe : just a thought : can we hack metrics in DroidSans to make a requested size 10 output size 9 glyph metrics ? :)
Or this would mess when used with KOReader's UI, where I have no issue with its monospace font (ie text editor).

May be this could be left to crengine, via a `crengine.font.monospace.scale = 90%` (no UI item for tuning that, but unhappy users could manually set some other value and restore 100%).
That is, if we don't find another monotype font that we would love and would support many languages.

**Comment by poire-z** [comment]
Not certain Droid Sans was designed to look alright near Noto Serif (dunno their history and which came before the other).
Guess we should give Noto Sans Mono a try (unless somebody is sure it's just DroidSans masquerading as some new font)
https://fonts.google.com/noto/specimen/Noto+Sans+Mono

**Comment by poire-z** [comment]
Not sure it looks really better (at least, not noticably smaller):
![image](https://user-images.githubusercontent.com/24273478/173664786-7dcbcf09-d9e6-4e9e-bd45-5dd79d189f64.png)

Noto Sans Mono has some char "more serif" (the `r`, `i`) than Droid Sans Mono.

**Comment by Frenzie** [comment]
> Not certain Droid Sans was designed to look alright near Noto Serif (dunno their history and which came before the other).

They're not *identical* but I think it's not misleading to say that the Noto font family is by and large the same thing as the Droid family except with more characters. (There are a few very noteworthy differences like the g of course.)

**Comment by NiLuJe** [comment]
> @NiLuJe : just a thought : can we hack metrics in DroidSans to make a requested size 10 output size 9 glyph metrics ? :) Or this would mess when used with KOReader's UI, where I have no issue with its monospace font (ie text editor).
> 
> May be this could be left to crengine, via a `crengine.font.monospace.scale = 90%` (no UI item for tuning that, but unhappy users could manually set some other value and restore 100%). That is, if we don't find another monotype font that we would love and would support many languages.
> 
> edit: but this would scale down any monospace font (even embedded ones that a publisher has hand picked to look nice... so, it would have to target only our fat DroidSans)

No idea, but, yeah, probably easier as a silent filter to the rendering size requested by CRe ;).

**Comment by mergen3107** [comment]
About making those monospace bits distinguishing from the rest of the text - how about applying style tweaks to make them bolder/lighter by 0.25? Or something like that, I mean maybe not necessarily a font change. However this probably won’t solve the issue in the OP about heights

**Comment by poire-z** [comment]
Using style tweaks would have some limited effects.
I mean, a user or a publisher can make things nice/nicer with styles, but what I'd like is just a snippet like this to look nice:
```html
<style>.MyCodeClass { font-family: monospace; }</style>
<p>Some normal text and <span style="font-family: monospace">some code</span>
and <span class="MyCodeClass">some other code</span> blah blah</p>
```
We could have style tweaks making `<pre>` and `<code>` smaller and bolder, but it will not catch all (as most style tweaks though).
Having a nice monospace font would solve all these cases.

(cre.cpp has `props->setString("styles.pre.font-face", "font-family: \"Droid Sans Mono\"");`, but this has no effect with KOReader - it's used by CoolReader strange/old UI style handling) - so the behaviour we get when there are multiple monospace font feels a bit undefined :/)

**Comment by poire-z** [comment]
Which "it" ? DroidSansMono ? Yes, it is.
This is  some (possibly old) Noto Sans Mono I have (100%, untouched, vs. Noto Serif):
![image](https://user-images.githubusercontent.com/24273478/174632318-fbc86030-4606-43ed-a014-c75b783a4721.png)
(its `r` is serif :) and its `g` loop is on the right).

**Comment by Frenzie** [comment]
I meant the size thing, which must be by design.

**Comment by Frenzie** [comment]
I thought Noto Serif was your example though?

**Comment by poire-z** [comment]
Yes, that's the screenshot in the first post.
It's indeed the same size, and possibly by design, and it is "acceptable" although I and @offset-torque think it looks better when a bit smaller as in https://github.com/koreader/koreader/issues/9211#issuecomment-1160556267.
But the "Palatino currently" in there is what many people will get with a fancier serif font, and I find it unacceptable.
So, we should do something. The x-height idea would target consistent "by design and acceptable - but not a little bit smaller" :), while 90% would target some non-consistent "a bit smaller with tall fonts, but acceptable with other fancier fonts" (which would still be fine by me).

**Comment by Frenzie** [comment]
If the x-height idea is attainable, that sounds good. :-)

**Comment by poire-z** [comment]
(Haven't yet thought about how/if we can do this x-height idea with monospace.)

But it had me thinking about whether "aligning x-heights" for the fallback fonts would be better that what we have.
Currently, we may have in a chain of fallbacks:
`Bitter (20) > NotoSansCJK (20) > NotoSansArabic(20) > FreeSans (20) > FreeSerif (20)`
and each font at size 20 can give glyphs that have a different visual height/weight.

By trying to find the size for the next font in the chain so that the x-height (the height of lowercase `x`) stay similar, we may get better looking text, and we can end up with:
`Bitter (20) > NotoSansCJK (18) > NotoSansArabic(20) > FreeSans (19) > FreeSerif (23)`
(FreeSerif is quite a lot smaller than FreeSans, so we would pick a larger font size than FreeSans to get glyphs with a similar look.)

I'm posting some screenshots, with bits of Greek, Chinese, Hebrew, Arabic - often picked from a fallback font, vs different main fonts:
KOReader current (no size tweak) | KOReader with fallback sizes tweaked:

Noto Serif:
![image](https://user-images.githubusercontent.com/24273478/174810836-3ddf65b2-5fcc-4184-a45b-a8d0151f4661.png)

Bitter:
![image](https://user-images.githubusercontent.com/24273478/174810582-cdb9d74c-3bb5-4ce1-814e-08889e394687.png)

Capita:
![image](https://user-images.githubusercontent.com/24273478/174809948-6076d133-de44-4a41-b7e5-2f06e03fb143.png)
(the greek character Pi happens to be found in the font, but not the other ones, so in "peri" the "p" is a bit smaller than the "eri" on the left side - and it looks fine on the right side).

Libre Baskerville at a larger font size:
![image](https://user-images.githubusercontent.com/24273478/174811988-74487c46-f969-4f0f-a66f-59b6cf6dfad3.png)

Sitka at a larger font size:
![image](https://user-images.githubusercontent.com/24273478/174812570-f8741410-9ec4-4aa2-82a6-3d0547b2cedb.png)

Volkorn at a smaller font size:
![image](https://user-images.githubusercontent.com/24273478/174812981-a5e4e220-5cfe-4589-a0b4-89e339101a69.png)
(the Hebrew got a bit fat...)

Any thought ? Preferences among the left or right renderings in the screenshots above ?

Pinging @ichnilatis-gr (for Greek), @yparitcher (for Hebrew) and @weijiuqiao (for Chinese) : how do you feel about the sizing of the script you're more familar with than me alongside words in latin script ? (the bits in your script probably don't mean anything, I must have typed them randomly :) but may be they are enough to get a visual feeling, even if subtle).
I don't read any of these scripts - even Greek, I usually just try to pick each letter and pronouce it to see if it sounds like a french word I know :) so I may end up focusing on these and notice the differences/mismatches in size may be more than a fluent reader.)
No idea about Chinese sizing (it sometimes become smaller in the right screenshots above).
About Hebrew, it gets "bigger" - which feels odd, but it may be that from the chars I picked, these look like lowercase `ooo` to me :) but they may in fact be bigger hebrew "squared" glyphs and should look the same size as uppercase latin.

**Comment by yparitcher** [comment]
>  About Hebrew, it gets "bigger" - which feels odd, but it may be that from the chars I picked, these look like lowercase `ooo` to me :) but they may in fact be bigger hebrew "squared" glyphs and should look the same size as uppercase latin.

1. I have a hebrew font set for the default in the folders that have hebrew books, so i never see the falbacks.
2. Hebrew does not have upper/lowercase letters. all letters take the full box (uppercase), אב‎ג‎ד‎ה‎ו‎ז‎ח‎ט‎כי‎‎םמ‎נסעפצ‎רשת
Diatrics will be above or below the box (The box sould be slightly smaller than line height to allow for this)
A few letters stick down from the baseline, ךןףץק and one letter sticks up, ל

**Comment by poire-z** [comment]
@yparitcher : thanks for the explanations.
But even if you don't/won't get to see our fallbacks in use, among the left|right screenshots above (click to see them larger if needed), is there any side that stand out as more correct/nicer looking/less odd than the other - or is it mixed feelings and/or no real need to have us care ?

**Comment by NiLuJe** [comment]
Slightly related query: what happens to differing line-heights in these scenario? (e.g., no rivers in the main font, and a fallback font with giant line-heights (hi, Noto)).

**Comment by poire-z** [comment]
line-height, strut, vertical-align... are all done from the metrics of the main font.
It's only when measuring widths (picking glyphs' advances) and drawing (picking grlyphs' bitmaps) that we go pick them into the chain of fallback fonts until found - and we draw them at the baseline/valign's `y` made out from the main font.
So, with a small main font with a small line-height, they will have no impact on the interline, they may just be drawn over the next or prev line if the fallback happens to have tall glyphs - by not much, that's the most extrem case I've found among my fonts, Baskervaldx, and it's still ok:
Current:
![image](https://user-images.githubusercontent.com/24273478/174889812-bd2e77c1-2278-4886-8bde-9fba897534b5.png)
With x-height aligning (it does not target at avoiding overlaps, but the reduction of size helps it a bit):
![image](https://user-images.githubusercontent.com/24273478/174890676-f490752c-c768-4f50-b2a2-c327d4118afb.png)

**Comment by weijiuqiao** [comment]
Chinese characters don't have x-height either, but most Chinese fonts have Latin alphabet inside, so the x-height of a font is mostly likely for the bundled letters. So in theory, since they are in the same font, their sizes should be designed to balance, then if they were to be resized to the same extent such that the bundled Latin letters have the same actual x-height value as the main font (so they would look great together), the Chinese part should also go better with the main font.

But in reality, I'm so used to Chinese characters being a little taller than surrounding Latin letters' cap-height and a little lower than their baseline that when they are smaller, they look a different font size. So sometimes the left side looks better, sometimes the right (but not significantly, and the serif/sans serif difference is also a distraction here).
